File: Assets/IFramework/HotFix/Lua/Resources/UI/LuaGroups.lua
local VVMGroup = require("UI.VVMGroup")
local Delegate = require("Base.Classes.Delegate")
---@type LuaGroups
local LuaGroups = class("LuaGroups")
local __CSGroups = StaticUsing("IFramework.Hotfix.Lua.LuaGroups")
function LuaGroups:ctor()
    self.onDispose = Handler(self, self.OnDispose)
    self.onSubscribe = Handler(self, self.OnSubscribe)
    self.onUnSubscribe = Handler(self, self.OnUnSubscribe)
    self.onFindPanel = Handler(self, self.OnFindPanel)
    self.onPress = Handler(self, self.OnPress)
    self.onTop = Handler(self, self.OnTop)
    self.onPop = Handler(self,self.OnPop)
    self.onShow = Handler(self, self.OnShow)
    self.onHide = Handler(self, self.OnHide)
    self.onPause = Handler(self, self.OnPause)
    self.onUnPause = Handler(self, self.OnUnPause)
    self.onClose = Handler(self, self.OnClose)
    self.groups = {}
end

---map: { { Name = "**",ViewType =require("****"), VMType=require("***")},}
--- @param map table 自动生成的
--- @return CS.IFramework.Hotfix.Lua.LuaGroups
function LuaGroups:SetMap(map)
    if map == nil then
        error("map could not be null ")
        return
    end
    self.CS_instance = __CSGroups()
    self.CS_instance:onDispose("+", self.onDispose)
    self.CS_instance:onSubscribe("+", self.onSubscribe)
    self.CS_instance:onUnSubscribe("+", self.onUnSubscribe)
    self.CS_instance:onFindPanel("+", self.onFindPanel)

    self.CS_instance:onPress("+", self.onPress)
    self.CS_instance:onPop("+", self.onPop)
    self.CS_instance:onTop("+", self.onTop)

    self.CS_instance:onShow("+", self.onShow)
    self.CS_instance:onHide("+", self.onHide)
    self.CS_instance:onPause("+", self.onPause)
    self.CS_instance:onUnPause("+", self.onUnPause)
    self.CS_instance:onClose("+", self.onClose)
    self.map = map
    return self.CS_instance
end

-- 以下方法私有

function LuaGroups:OnDispose()
    for i, group in pairs(self.groups) do
        group.view:OnClear()
        group:Dispose()
    end

    self.CS_instance:onDispose("-", self.onDispose)
    self.CS_instance:onSubscribe("-", self.onSubscribe)
    self.CS_instance:onUnSubscribe("-", self.onUnSubscribe)
    self.CS_instance:onFindPanel("-", self.onFindPanel)
    self.CS_instance:onPress("-", self.onPress)
    self.CS_instance:onPop("-", self.onPop)
    self.CS_instance:onTop("-", self.onTop)
    self.CS_instance:onShow("-", self.onShow)
    self.CS_instance:onHide("-", self.onHide)
    self.CS_instance:onPause("-", self.onPause)
    self.CS_instance:onUnPause("-", self.onUnPause)
    self.CS_instance:onClose("-", self.onClose)
    self.groups = nil
    self.CS_instance = nil
    self.map = nil
end
function LuaGroups:FindGroup(name)
    return rawget(self.groups, name)
end
function LuaGroups:OnSubscribe(panel)
    local name = panel.name
    if rawget(self.groups, panel.name) ~= nil then
        print("same name with panel  " .. panel.name)
        return false
    end
    local vvmType

    for i, v in pairs(self.map) do
        if v.Name == name then
            vvmType = v
            break
        end
    end
    if (vvmType == nil) then
        error("not find vvm type with Name :" .. name)
        return false
    end

    local message = Delegate()
    local viewModel = vvmType.VMType(message)
    local view = vvmType.ViewType(message, viewModel, panel)
    local vvmGroup = VVMGroup(panel, view, viewModel, message)
    rawset(self.groups, panel.name, vvmGroup)
end
function LuaGroups:OnUnSubscribe(panel)
    local group = rawget(self.groups, panel.name)
    if group ~= nil then
        group.view:OnClear()
        group:Dispose()
        rawset(self.groups, panel.name, nil)
        return true
    end
    return false
end

function LuaGroups:OnFindPanel(name)
    local group = self:FindGroup(name)
    if group ~= nil then
        return group.panel
    end
    return nil
end
function LuaGroups:OnTop(name, arg)
    self:FindGroup(name).view:OnTop(arg)
end
function LuaGroups:OnPress(name, arg)
    self:FindGroup(name).view:OnPress(arg)
end
function LuaGroups:OnPop(name, arg)
    self:FindGroup(name).view:OnPop(arg)
end

function LuaGroups:OnShow(name)
    self:FindGroup(name).view:OnShow()
end

function LuaGroups:OnHide(name)
    self:FindGroup(name).view:OnHide()
end

function LuaGroups:OnPause(name)
    self:FindGroup(name).view:OnPause()
end

function LuaGroups:OnUnPause(name)
    self:FindGroup(name).view:OnUnPause()
end

function LuaGroups:OnClose(name)
    self:FindGroup(name).view:OnClose()
end
return LuaGroups

File: Assets/IFramework/HotFix/Lua/Scripts/UI/LuaGroups.cs
/*********************************************************************************
 *Author:         OnClick
 *Version:        0.0.2.318
 *UnityVersion:   2018.4.17f1
 *Date:           2020-06-01
 *Description:    IFramework
 *History:        2018.11--
*********************************************************************************/

using System;
using IFramework.UI;

namespace IFramework.Hotfix.Lua
{
    public class LuaGroups : IGroups
    {
        public event Action onDispose;
        public event Func<string, UIPanel> onFindPanel;
        public event Action<string, UIEventArgs> onPop;
        public event Action<string, UIEventArgs> onPress;
        public event Action<string, UIEventArgs> onTop;

        public event Func<UIPanel, bool> onSubscribe;
        public event Func<UIPanel, bool> onUnSubscribe;
        public event Action<string> onShow;
        public event Action<string> onHide;
        public event Action<string> onPause;
        public event Action<string> onUnPause;
        public event Action<string> onClose;

        void IDisposable.Dispose()
        {
            if (onDispose != null)
            {
                onDispose();
            }
            onDispose = null;
            onFindPanel = null;
            onSubscribe = null;
            onUnSubscribe = null;
        }
        void IGroups.OnPop(string name, UIEventArgs arg)
        {
            if (onPop != null)
            {
                onPop(name, arg);
            }
        }

        void IGroups.OnPress(string name, UIEventArgs arg)
        {
            if (onPress != null)
            {
                onPress(name, arg);
            }
        }
        void IGroups.OnTop(string name, UIEventArgs arg)
        {
            if (onTop != null)
            {
                onTop(name, arg);
            }
        }


        UIPanel IGroups.FindPanel(string name)
        {
            if (onFindPanel != null)
            {
                return onFindPanel(name);
            }
            return null;
        }





        void IGroups.OnClose(string name)
        {
            if (onClose != null)
            {
                onClose.Invoke(name);
            }
        }

        void IGroups.OnHide(string name)
        {
            if (onHide != null)
            {
                onHide.Invoke(name);
            }
        }

        void IGroups.OnPause(string name)
        {
            if (onPause != null)
            {
                onPause.Invoke(name);
            }
        }



        void IGroups.OnShow(string name)
        {
            if (onShow != null)
            {
                onShow.Invoke(name);
            }
        }


        void IGroups.OnUnPause(string name)
        {
            if (onUnPause != null)
            {
                onUnPause.Invoke(name);
            }
        }

        bool IGroups.Subscribe(UIPanel panel)
        {

            if (onSubscribe != null)
            {
                return onSubscribe(panel);
            }
            return false;
        }

        bool IGroups.UnSubscribe(UIPanel panel)
        {
            if (onUnSubscribe != null)
            {
                return onUnSubscribe(panel);
            }

            return false;
        }

        void IGroups.Update()
        {

        }
    }

}
